pylint代码静态检测结果

主题：根据 pylint 的检测结果优化相应代码。

执行命令 *pylint borax* ，结果如下：

```
************* Module borax.choices
borax\choices.py:76:56: E1101: Function 'fields' has no 'items' member (no-member)
borax\choices.py:80:21: E1101: Function 'fields' has no 'keys' member (no-member)
borax\choices.py:84:44: E1133: Non-iterable value cls.choices is used in an iterating context (not-an-iterable)
borax\choices.py:88:48: E1133: Non-iterable value cls.choices is used in an iterating context (not-an-iterable)
borax\choices.py:96:49: E1133: Non-iterable value cls.choices is used in an iterating context (not-an-iterable)
borax\choices.py:96:0: R1721: Unnecessary use of a comprehension (unnecessary-comprehension)
borax\choices.py:98:4: C0203: Metaclass method __contains__ should have 'cls' as first argument (bad-mcs-method-argument)
borax\choices.py:99:23: E1135: Value 'self.values' doesn't support membership test (unsupported-membership-test)
borax\choices.py:101:4: C0203: Metaclass method __iter__ should have 'cls' as first argument (bad-mcs-method-argument)
borax\choices.py:102:20: E1133: Non-iterable value self.choices is used in an iterating context (not-an-iterable)
borax\choices.py:105:4: C0203: Metaclass method __len__ should have 'cls' as first argument (bad-mcs-method-argument)
borax\choices.py:110:24: E1135: Value 'cls.display_lookup' doesn't support membership test (unsupported-membership-test)
borax\choices.py:113:15: E1101: Function 'display_lookup' has no 'get' member (no-member)
borax\choices.py:116:0: R0903: Too few public methods (0/2) (too-few-public-methods)
************* Module borax.constants
borax\constants.py:5:0: C0301: Line too long (123/120) (line-too-long)
borax\constants.py:8:0: R0903: Too few public methods (0/2) (too-few-public-methods)
************* Module borax.htmls
borax\htmls.py:51:4: R1705: Unnecessary "else" after "return" (no-else-return)
************* Module borax.numbers
borax\numbers.py:64:0: R0903: Too few public methods (1/2) (too-few-public-methods)
************* Module borax.utils
borax\utils.py:108:0: R1707: Disallow trailing comma tuple (trailing-comma-tuple)
borax\utils.py:8:4: R1705: Unnecessary "else" after "return" (no-else-return)
borax\utils.py:103:4: R1705: Unnecessary "elif" after "return" (no-else-return)
************* Module borax.calendars.festivals
borax\calendars\festivals.py:188:0: C0325: Unnecessary parens after 'not' keyword (superfluous-parens)
borax\calendars\festivals.py:25:0: W0613: Unused argument 'args' (unused-argument)
borax\calendars\festivals.py:37:15: E1101: Instance of 'DateSchema' has no 'year' member (no-member)
borax\calendars\festivals.py:46:65: W0212: Access to a protected member __key of a client class (protected-access)
borax\calendars\festivals.py:73:23: E1101: Instance of 'DateSchema' has no 'year' member (no-member)
borax\calendars\festivals.py:83:8: R1705: Unnecessary "else" after "return" (no-else-return)
borax\calendars\festivals.py:141:4: R0913: Too many arguments (6/5) (too-many-arguments)
borax\calendars\festivals.py:287:4: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
************* Module borax.calendars.lunardate
borax\calendars\lunardate.py:78:4: W0612: Unused variable 'leap_month' (unused-variable)
borax\calendars\lunardate.py:135:8: W0120: Else clause on loop without a break statement (useless-else-on-loop)
borax\calendars\lunardate.py:143:35: R1714: Consider merging these comparisons with "in" to 'month in (0, leap_month)' (consider-using-in)
borax\calendars\lunardate.py:168:8: R1705: Unnecessary "else" after "return" (no-else-return)
borax\calendars\lunardate.py:189:15: W0631: Using possibly undefined loop variable '_month' (undefined-loop-variable)
borax\calendars\lunardate.py:189:36: W0631: Using possibly undefined loop variable '_leap' (undefined-loop-variable)
borax\calendars\lunardate.py:199:28: W0631: Using possibly undefined loop variable 'idx' (undefined-loop-variable)
borax\calendars\lunardate.py:201:27: W0631: Using possibly undefined loop variable 'idx' (undefined-loop-variable)
borax\calendars\lunardate.py:212:16: R1705: Unnecessary "else" after "return" (no-else-return)
borax\calendars\lunardate.py:351:12: R1705: Unnecessary "else" after "return" (no-else-return)
borax\calendars\lunardate.py:364:0: R0902: Too many instance attributes (9/7) (too-many-instance-attributes)
borax\calendars\lunardate.py:456:12: R1705: Unnecessary "else" after "return" (no-else-return)
borax\calendars\lunardate.py:544:8: R1705: Unnecessary "elif" after "return" (no-else-return)
borax\calendars\lunardate.py:553:4: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
borax\calendars\lunardate.py:570:12: W0707: Consider explicitly re-raising using the 'from' keyword (raise-missing-from)
borax\calendars\lunardate.py:588:65: W0212: Access to a protected member __key of a client class (protected-access)
borax\calendars\lunardate.py:364:0: R0904: Too many public methods (30/20) (too-many-public-methods)
borax\calendars\lunardate.py:653:12: R1705: Unnecessary "else" after "return" (no-else-return)
borax\calendars\lunardate.py:660:4: R0201: Method could be a function (no-self-use)
borax\calendars\lunardate.py:663:4: R0201: Method could be a function (no-self-use)
borax\calendars\lunardate.py:667:8: R1705: Unnecessary "else" after "return" (no-else-return)
borax\calendars\lunardate.py:666:4: R0201: Method could be a function (no-self-use)
borax\calendars\lunardate.py:672:4: R0201: Method could be a function (no-self-use)
borax\calendars\lunardate.py:675:4: R0201: Method could be a function (no-self-use)
borax\calendars\lunardate.py:10:0: C0411: standard import "from typing import Optional, Iterator, Tuple" should be placed before "from .store import EncoderMixin, f_year, f_month, f_day, f_leap" (wrong-import-order)
************* Module borax.calendars.store
borax\calendars\store.py:19:12: R1705: Unnecessary "else" after "return" (no-else-return)
borax\calendars\store.py:27:8: R1705: Unnecessary "else" after "return" (no-else-return)
************* Module borax.counters.daily
borax\counters\daily.py:7:2: W0511: TODO Use Counter (fixme)
borax\counters\daily.py:26:8: R1705: Unnecessary "else" after "return" (no-else-return)
************* Module borax.counters.serial_pool
borax\counters\serial_pool.py:69:4: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
borax\counters\serial_pool.py:75:0: R0903: Too few public methods (0/2) (too-few-public-methods)
borax\counters\serial_pool.py:84:4: R0913: Too many arguments (6/5) (too-many-arguments)
************* Module borax.datasets.fetch
borax\datasets\fetch.py:11:0: R0903: Too few public methods (0/2) (too-few-public-methods)
borax\datasets\fetch.py:40:8: R1705: Unnecessary "else" after "return" (no-else-return)
borax\datasets\fetch.py:64:4: R1705: Unnecessary "else" after "return" (no-else-return)
borax\datasets\fetch.py:76:11: R1717: Consider using a dictionary comprehension (consider-using-dict-comprehension)
************* Module borax.datasets.join_
borax\datasets\join_.py:41:8: R1705: Unnecessary "elif" after "return" (no-else-return)
borax\datasets\join_.py:59:8: R1705: Unnecessary "elif" after "return" (no-else-return)
borax\datasets\join_.py:93:12: W0120: Else clause on loop without a break statement (useless-else-on-loop)
borax\datasets\join_.py:143:0: R0913: Too many arguments (6/5) (too-many-arguments)
************* Module borax.patterns.lazy
borax\patterns\lazy.py:15:11: W0212: Access to a protected member _wrapped of a client class (protected-access)
borax\patterns\lazy.py:16:12: W0212: Access to a protected member _setup of a client class (protected-access)
borax\patterns\lazy.py:17:20: W0212: Access to a protected member _wrapped of a client class (protected-access)
************* Module borax.patterns.proxy
borax\patterns\proxy.py:26:4: E0213: Method should have "self" as first argument (no-self-argument)
borax\patterns\proxy.py:46:15: R1725: Consider using Python 3 style super() without arguments (super-with-arguments)
************* Module borax.serialize.bjson
borax\serialize\bjson.py:11:0: R0903: Too few public methods (1/2) (too-few-public-methods)
************* Module borax.serialize.cjson
borax\serialize\cjson.py:22:0: E1111: Assigning result of a function call, where the function has no return (assignment-from-no-return)
borax\serialize\cjson.py:22:21: E1120: No value for argument 'cls' in function call (no-value-for-parameter)
borax\serialize\cjson.py:22:21: E1120: No value for argument 'type' in function call (no-value-for-parameter)
************* Module borax.structures.dictionary
borax\structures\dictionary.py:9:12: W0707: Consider explicitly re-raising using the 'from' keyword (raise-missing-from)
************* Module borax.structures.percentage
borax\structures\percentage.py:34:8: R1705: Unnecessary "else" after "return" (no-else-return)
************* Module borax.structures.tree
borax\structures\tree.py:32:0: R0914: Too many local variables (17/15) (too-many-locals)
borax\structures\tree.py:66:8: R1715: Consider using dict.get for getting values from a dict if a key is present or a default if not (consider-using-get)
************* Module borax.ui.widgets
borax\ui\widgets.py:6:0: R0901: Too many ancestors (8/7) (too-many-ancestors)
borax\ui\widgets.py:30:0: R0901: Too many ancestors (9/7) (too-many-ancestors)

------------------------------------------------------------------

Your code has been rated at 9.12/10 (previous run: 9.12/10, +0.00)




```

--- a/.pylintrc
+++ b/.pylintrc
@@ -1,0 +1,594 @@
+[MASTER]
+
+# A comma-separated list of package or module names from where C extensions may
+# be loaded. Extensions are loading into the active Python interpreter and may
+# run arbitrary code.
+extension-pkg-whitelist=
+
+# Specify a score threshold to be exceeded before program exits with error.
+fail-under=10.0
+
+# Add files or directories to the blacklist. They should be base names, not
+# paths.
+ignore=CVS
+
+# Add files or directories matching the regex patterns to the blacklist. The
+# regex matches against base names, not paths.
+ignore-patterns=
+
+# Python code to execute, usually for sys.path manipulation such as
+# pygtk.require().
+#init-hook=
+
+# Use multiple processes to speed up Pylint. Specifying 0 will auto-detect the
+# number of processors available to use.
+jobs=1
+
+# Control the amount of potential inferred values when inferring a single
+# object. This can help the performance when dealing with large functions or
+# complex, nested conditions.
+limit-inference-results=100
+
+# List of plugins (as comma separated values of python module names) to load,
+# usually to register additional checkers.
+load-plugins=
+
+# Pickle collected data for later comparisons.
+persistent=yes
+
+# When enabled, pylint would attempt to guess common misconfiguration and emit
+# user-friendly hints instead of false-positive error messages.
+suggestion-mode=yes
+
+# Allow loading of arbitrary C extensions. Extensions are imported into the
+# active Python interpreter and may run arbitrary code.
+unsafe-load-any-extension=no
+
+
+[MESSAGES CONTROL]
+
+# Only show warnings with the listed confidence levels. Leave empty to show
+# all. Valid levels: HIGH, INFERENCE, INFERENCE_FAILURE, UNDEFINED.
+confidence=
+
+# Disable the message, report, category or checker with the given id(s). You
+# can either give multiple identifiers separated by comma (,) or put this
+# option multiple times (only on the command line, not in the configuration
+# file where it should appear only once). You can also use "--disable=all" to
+# disable everything first and then reenable specific checks. For example, if
+# you want to run only the similarities checker, you can use "--disable=all
+# --enable=similarities". If you want to run only the classes checker, but have
+# no Warning level messages displayed, use "--disable=all --enable=classes
+# --disable=W".
+disable=print-statement,
+        parameter-unpacking,
+        unpacking-in-except,
+        old-raise-syntax,
+        backtick,
+        long-suffix,
+        old-ne-operator,
+        old-octal-literal,
+        import-star-module-level,
+        non-ascii-bytes-literal,
+        raw-checker-failed,
+        bad-inline-option,
+        locally-disabled,
+        file-ignored,
+        suppressed-message,
+        useless-suppression,
+        deprecated-pragma,
+        use-symbolic-message-instead,
+        apply-builtin,
+        basestring-builtin,
+        buffer-builtin,
+        cmp-builtin,
+        coerce-builtin,
+        execfile-builtin,
+        file-builtin,
+        long-builtin,
+        raw_input-builtin,
+        reduce-builtin,
+        standarderror-builtin,
+        unicode-builtin,
+        xrange-builtin,
+        coerce-method,
+        delslice-method,
+        getslice-method,
+        setslice-method,
+        no-absolute-import,
+        old-division,
+        dict-iter-method,
+        dict-view-method,
+        next-method-called,
+        metaclass-assignment,
+        indexing-exception,
+        raising-string,
+        reload-builtin,
+        oct-method,
+        hex-method,
+        nonzero-method,
+        cmp-method,
+        input-builtin,
+        round-builtin,
+        intern-builtin,
+        unichr-builtin,
+        map-builtin-not-iterating,
+        zip-builtin-not-iterating,
+        range-builtin-not-iterating,
+        filter-builtin-not-iterating,
+        using-cmp-argument,
+        eq-without-hash,
+        div-method,
+        idiv-method,
+        rdiv-method,
+        exception-message-attribute,
+        invalid-str-codec,
+        sys-max-int,
+        bad-python3-import,
+        deprecated-string-function,
+        deprecated-str-translate-call,
+        deprecated-itertools-function,
+        deprecated-types-field,
+        next-method-defined,
+        dict-items-not-iterating,
+        dict-keys-not-iterating,
+        dict-values-not-iterating,
+        deprecated-operator-function,
+        deprecated-urllib-function,
+        xreadlines-attribute,
+        deprecated-sys-function,
+        exception-escape,
+        comprehension-escape,
+		missing-function-docstring,
+		missing-class-docstring,
+		missing-module-docstring,
+		bad-classmethod-argument,
+		invalid-name,
+		no-else-return
+
+# Enable the message, report, category or checker with the given id(s). You can
+# either give multiple identifier separated by comma (,) or put this option
+# multiple time (only on the command line, not in the configuration file where
+# it should appear only once). See also the "--disable" option for examples.
+enable=c-extension-no-member
+
+
+[REPORTS]
+
+# Python expression which should return a score less than or equal to 10. You
+# have access to the variables 'error', 'warning', 'refactor', and 'convention'
+# which contain the number of messages in each category, as well as 'statement'
+# which is the total number of statements analyzed. This score is used by the
+# global evaluation report (RP0004).
+evaluation=10.0 - ((float(5 * error + warning + refactor + convention) / statement) * 10)
+
+# Template used to display messages. This is a python new-style format string
+# used to format the message information. See doc for all details.
+#msg-template=
+
+# Set the output format. Available formats are text, parseable, colorized, json
+# and msvs (visual studio). You can also give a reporter class, e.g.
+# mypackage.mymodule.MyReporterClass.
+output-format=text
+
+# Tells whether to display a full report or only the messages.
+reports=no
+
+# Activate the evaluation score.
+score=yes
+
+
+[REFACTORING]
+
+# Maximum number of nested blocks for function / method body
+max-nested-blocks=5
+
+# Complete name of functions that never returns. When checking for
+# inconsistent-return-statements if a never returning function is called then
+# it will be considered as an explicit return statement and no message will be
+# printed.
+never-returning-functions=sys.exit
+
+
+[BASIC]
+
+# Naming style matching correct argument names.
+argument-naming-style=snake_case
+
+# Regular expression matching correct argument names. Overrides argument-
+# naming-style.
+#argument-rgx=
+
+# Naming style matching correct attribute names.
+attr-naming-style=snake_case
+
+# Regular expression matching correct attribute names. Overrides attr-naming-
+# style.
+#attr-rgx=
+
+# Bad variable names which should always be refused, separated by a comma.
+bad-names=foo,
+          bar,
+          baz,
+          toto,
+          tutu,
+          tata
+
+# Bad variable names regexes, separated by a comma. If names match any regex,
+# they will always be refused
+bad-names-rgxs=
+
+# Naming style matching correct class attribute names.
+class-attribute-naming-style=any
+
+# Regular expression matching correct class attribute names. Overrides class-
+# attribute-naming-style.
+#class-attribute-rgx=
+
+# Naming style matching correct class names.
+class-naming-style=PascalCase
+
+# Regular expression matching correct class names. Overrides class-naming-
+# style.
+#class-rgx=
+
+# Naming style matching correct constant names.
+const-naming-style=UPPER_CASE
+
+# Regular expression matching correct constant names. Overrides const-naming-
+# style.
+#const-rgx=
+
+# Minimum line length for functions/classes that require docstrings, shorter
+# ones are exempt.
+docstring-min-length=-1
+
+# Naming style matching correct function names.
+function-naming-style=snake_case
+
+# Regular expression matching correct function names. Overrides function-
+# naming-style.
+#function-rgx=
+
+# Good variable names which should always be accepted, separated by a comma.
+good-names=i,
+           j,
+           k,
+           ex,
+           Run,
+           _
+
+# Good variable names regexes, separated by a comma. If names match any regex,
+# they will always be accepted
+good-names-rgxs=
+
+# Include a hint for the correct naming format with invalid-name.
+include-naming-hint=no
+
+# Naming style matching correct inline iteration names.
+inlinevar-naming-style=any
+
+# Regular expression matching correct inline iteration names. Overrides
+# inlinevar-naming-style.
+#inlinevar-rgx=
+
+# Naming style matching correct method names.
+method-naming-style=snake_case
+
+# Regular expression matching correct method names. Overrides method-naming-
+# style.
+#method-rgx=
+
+# Naming style matching correct module names.
+module-naming-style=snake_case
+
+# Regular expression matching correct module names. Overrides module-naming-
+# style.
+#module-rgx=
+
+# Colon-delimited sets of names that determine each other's naming style when
+# the name regexes allow several styles.
+name-group=
+
+# Regular expression which should only match function or class names that do
+# not require a docstring.
+no-docstring-rgx=^_
+
+# List of decorators that produce properties, such as abc.abstractproperty. Add
+# to this list to register other decorators that produce valid properties.
+# These decorators are taken in consideration only for invalid-name.
+property-classes=abc.abstractproperty
+
+# Naming style matching correct variable names.
+variable-naming-style=snake_case
+
+# Regular expression matching correct variable names. Overrides variable-
+# naming-style.
+#variable-rgx=
+
+
+[FORMAT]
+
+# Expected format of line ending, e.g. empty (any line ending), LF or CRLF.
+expected-line-ending-format=
+
+# Regexp for a line that is allowed to be longer than the limit.
+ignore-long-lines=^\s*(# )?<?https?://\S+>?$
+
+# Number of spaces of indent required inside a hanging or continued line.
+indent-after-paren=4
+
+# String used as indentation unit. This is usually "    " (4 spaces) or "\t" (1
+# tab).
+indent-string='    '
+
+# Maximum number of characters on a single line.
+max-line-length=120
+
+# Maximum number of lines in a module.
+max-module-lines=1000
+
+# Allow the body of a class to be on the same line as the declaration if body
+# contains single statement.
+single-line-class-stmt=no
+
+# Allow the body of an if to be on the same line as the test if there is no
+# else.
+single-line-if-stmt=no
+
+
+[LOGGING]
+
+# The type of string formatting that logging methods do. `old` means using %
+# formatting, `new` is for `{}` formatting.
+logging-format-style=old
+
+# Logging modules to check that the string format arguments are in logging
+# function parameter format.
+logging-modules=logging
+
+
+[MISCELLANEOUS]
+
+# List of note tags to take in consideration, separated by a comma.
+notes=FIXME,
+      XXX,
+      TODO
+
+# Regular expression of note tags to take in consideration.
+#notes-rgx=
+
+
+[SIMILARITIES]
+
+# Ignore comments when computing similarities.
+ignore-comments=yes
+
+# Ignore docstrings when computing similarities.
+ignore-docstrings=yes
+
+# Ignore imports when computing similarities.
+ignore-imports=no
+
+# Minimum lines number of a similarity.
+min-similarity-lines=4
+
+
+[SPELLING]
+
+# Limits count of emitted suggestions for spelling mistakes.
+max-spelling-suggestions=4
+
+# Spelling dictionary name. Available dictionaries: none. To make it work,
+# install the python-enchant package.
+spelling-dict=
+
+# List of comma separated words that should not be checked.
+spelling-ignore-words=
+
+# A path to a file that contains the private dictionary; one word per line.
+spelling-private-dict-file=
+
+# Tells whether to store unknown words to the private dictionary (see the
+# --spelling-private-dict-file option) instead of raising a message.
+spelling-store-unknown-words=no
+
+
+[STRING]
+
+# This flag controls whether inconsistent-quotes generates a warning when the
+# character used as a quote delimiter is used inconsistently within a module.
+check-quote-consistency=no
+
+# This flag controls whether the implicit-str-concat should generate a warning
+# on implicit string concatenation in sequences defined over several lines.
+check-str-concat-over-line-jumps=no
+
+
+[TYPECHECK]
+
+# List of decorators that produce context managers, such as
+# contextlib.contextmanager. Add to this list to register other decorators that
+# produce valid context managers.
+contextmanager-decorators=contextlib.contextmanager
+
+# List of members which are set dynamically and missed by pylint inference
+# system, and so shouldn't trigger E1101 when accessed. Python regular
+# expressions are accepted.
+generated-members=
+
+# Tells whether missing members accessed in mixin class should be ignored. A
+# mixin class is detected if its name ends with "mixin" (case insensitive).
+ignore-mixin-members=yes
+
+# Tells whether to warn about missing members when the owner of the attribute
+# is inferred to be None.
+ignore-none=yes
+
+# This flag controls whether pylint should warn about no-member and similar
+# checks whenever an opaque object is returned when inferring. The inference
+# can return multiple potential results while evaluating a Python object, but
+# some branches might not be evaluated, which results in partial inference. In
+# that case, it might be useful to still emit no-member and other checks for
+# the rest of the inferred objects.
+ignore-on-opaque-inference=yes
+
+# List of class names for which member attributes should not be checked (useful
+# for classes with dynamically set attributes). This supports the use of
+# qualified names.
+ignored-classes=optparse.Values,thread._local,_thread._local
+
+# List of module names for which member attributes should not be checked
+# (useful for modules/projects where namespaces are manipulated during runtime
+# and thus existing member attributes cannot be deduced by static analysis). It
+# supports qualified module names, as well as Unix pattern matching.
+ignored-modules=
+
+# Show a hint with possible names when a member name was not found. The aspect
+# of finding the hint is based on edit distance.
+missing-member-hint=yes
+
+# The minimum edit distance a name should have in order to be considered a
+# similar match for a missing member name.
+missing-member-hint-distance=1
+
+# The total number of similar names that should be taken in consideration when
+# showing a hint for a missing member.
+missing-member-max-choices=1
+
+# List of decorators that change the signature of a decorated function.
+signature-mutators=
+
+
+[VARIABLES]
+
+# List of additional names supposed to be defined in builtins. Remember that
+# you should avoid defining new builtins when possible.
+additional-builtins=
+
+# Tells whether unused global variables should be treated as a violation.
+allow-global-unused-variables=yes
+
+# List of strings which can identify a callback function by name. A callback
+# name must start or end with one of those strings.
+callbacks=cb_,
+          _cb
+
+# A regular expression matching the name of dummy variables (i.e. expected to
+# not be used).
+dummy-variables-rgx=_+$|(_[a-zA-Z0-9_]*[a-zA-Z0-9]+?$)|dummy|^ignored_|^unused_
+
+# Argument names that match this expression will be ignored. Default to name
+# with leading underscore.
+ignored-argument-names=_.*|^ignored_|^unused_
+
+# Tells whether we should check for unused import in __init__ files.
+init-import=no
+
+# List of qualified module names which can have objects that can redefine
+# builtins.
+redefining-builtins-modules=six.moves,past.builtins,future.builtins,builtins,io
+
+
+[CLASSES]
+
+# List of method names used to declare (i.e. assign) instance attributes.
+defining-attr-methods=__init__,
+                      __new__,
+                      setUp,
+                      __post_init__
+
+# List of member names, which should be excluded from the protected access
+# warning.
+exclude-protected=_asdict,
+                  _fields,
+                  _replace,
+                  _source,
+                  _make
+
+# List of valid names for the first argument in a class method.
+valid-classmethod-first-arg=cls
+
+# List of valid names for the first argument in a metaclass class method.
+valid-metaclass-classmethod-first-arg=cls
+
+
+[DESIGN]
+
+# Maximum number of arguments for function / method.
+max-args=5
+
+# Maximum number of attributes for a class (see R0902).
+max-attributes=7
+
+# Maximum number of boolean expressions in an if statement (see R0916).
+max-bool-expr=5
+
+# Maximum number of branch for function / method body.
+max-branches=12
+
+# Maximum number of locals for function / method body.
+max-locals=15
+
+# Maximum number of parents for a class (see R0901).
+max-parents=7
+
+# Maximum number of public methods for a class (see R0904).
+max-public-methods=20
+
+# Maximum number of return / yield for function / method body.
+max-returns=6
+
+# Maximum number of statements in function / method body.
+max-statements=50
+
+# Minimum number of public methods for a class (see R0903).
+min-public-methods=2
+
+
+[IMPORTS]
+
+# List of modules that can be imported at any level, not just the top level
+# one.
+allow-any-import-level=
+
+# Allow wildcard imports from modules that define __all__.
+allow-wildcard-with-all=no
+
+# Analyse import fallback blocks. This can be used to support both Python 2 and
+# 3 compatible code, which means that the block might have code that exists
+# only in one or another interpreter, leading to false positives when analysed.
+analyse-fallback-blocks=no
+
+# Deprecated modules which should not be used, separated by a comma.
+deprecated-modules=optparse,tkinter.tix
+
+# Create a graph of external dependencies in the given file (report RP0402 must
+# not be disabled).
+ext-import-graph=
+
+# Create a graph of every (i.e. internal and external) dependencies in the
+# given file (report RP0402 must not be disabled).
+import-graph=
+
+# Create a graph of internal dependencies in the given file (report RP0402 must
+# not be disabled).
+int-import-graph=
+
+# Force import order to recognize a module as part of the standard
+# compatibility libraries.
+known-standard-library=
+
+# Force import order to recognize a module as part of a third party library.
+known-third-party=enchant
+
+# Couples of modules and preferred modules, separated by a comma.
+preferred-modules=
+
+
+[EXCEPTIONS]
+
+# Exceptions that will emit a warning when being caught. Defaults to
+# "BaseException, Exception".
+overgeneral-exceptions=BaseException,
+                       Exception

--- a/borax/calendars/lunardate.py
+++ b/borax/calendars/lunardate.py
@@ -75,7 +75,7 @@ def _parse_leap(year_info):
 def parse_year_days(year_info):
     """Parse year days from a year info.
     """
-    leap_month, leap_days = _parse_leap(year_info)
+    _, leap_days = _parse_leap(year_info)
     res = leap_days
     for month in range(1, 13):
         res += (year_info >> (16 - month)) % 2 + 29
@@ -553,6 +553,7 @@ class LunarDate(EncoderMixin):
     def __rsub__(self, other):
         if isinstance(other, datetime.date):
             return other - self.to_solar_date()
+        raise TypeError
 
     def __add__(self, other):
         if isinstance(other, datetime.timedelta):
@@ -566,8 +567,8 @@ class LunarDate(EncoderMixin):
     def __lt__(self, other):
         try:
             return self - other < datetime.timedelta(0)
-        except TypeError:
-            raise TypeError("can't compare LunarDate to %s" % (type(other).__name__,))
+        except TypeError as ex:
+            raise TypeError("can't compare LunarDate to %s" % (type(other).__name__,)) from ex
 
     def __le__(self, other):
         return self < other or self == other

--- a/borax/htmls.py
+++ b/borax/htmls.py
@@ -50,5 +50,4 @@ def html_params(**kwargs) -> str:
 def html_tag(tag_name, content=None, **kwargs):
     if content:
         return HTMLString('<{0} {1}>{2}</{0}>'.format(tag_name, html_params(**kwargs), content))
-    else:
-        return HTMLString('<{0} {1}>'.format(tag_name, html_params(**kwargs)))
+    return HTMLString('<{0} {1}>'.format(tag_name, html_params(**kwargs)))

--- a/borax/patterns/proxy.py
+++ b/borax/patterns/proxy.py
@@ -43,4 +43,4 @@ class Proxy:
     def __setattr__(self, attr, value):
         if attr not in self.__slots__:
             raise AttributeError('Cannot set attribute on proxy.')
-        return super(Proxy, self).__setattr__(attr, value)
+        return super().__setattr__(attr, value)

--- a/borax/structures/tree.py
+++ b/borax/structures/tree.py
@@ -61,12 +61,7 @@ def pll2cnl(
 
         node_id = _get(item, id_field)
         parent_id = _get(item, parent_field)
-        # create current node if necessary
-        # Create or update
-        if node_id in nodes:
-            node = nodes[node_id]
-        else:
-            node = {id_field: node_id}
+        node = nodes.get(node_id, {id_field: node_id})
         kwargs_data = _parse_extra_data(
             item,
             flat_fields=flat_fields,

